File: deepseek_infra/infra/rag/document_preparation.py
"""Deterministic, side-effect-free RAG document preparation.

The module accepts text that Python has already extracted from an uploaded
file.  It normalizes and chunks that text, but deliberately has no access to
uploads, paths, OCR, embeddings, caches, databases, or indexes.
"""

from __future__ import annotations

import hashlib
import json
import logging
import math
import os
import time
from dataclasses import dataclass
from typing import Any

RAG_DOCUMENT_MAX_CHARACTERS = 8_000_000
RAG_DOCUMENT_MAX_REQUEST_BYTES = 40_000_000
RAG_DOCUMENT_MAX_NESTING = 24
RAG_DOCUMENT_MAX_ID_CHARACTERS = 256
RAG_DOCUMENT_MAX_CHUNK_CHARACTERS = 1_000_000
DEFAULT_CHUNK_CHARACTERS = 6000
DEFAULT_CHUNK_OVERLAP = 400

_ALLOWED_TOP_LEVEL = frozenset({"documentId", "text", "metadata", "chunking"})
_ALLOWED_CHUNKING = frozenset({"chunkChars", "chunkOverlap"})
_ALLOWED_METADATA = frozenset({"displayName", "sourceType", "kind"})
_SENSITIVE_KEY_PARTS = (
    "absolutepath",
    "temporarypath",
    "uploadpath",
    "cachepath",
    "authorization",
    "apikey",
    "token",
    "rawfilebytes",
    "databaselocation",
    "workspacesecret",
    "filesystempath",
)

logger = logging.getLogger("deepseek_infra.rag.document_preparation")


@dataclass(frozen=True, slots=True)
class RagDocumentPreparationDecision:
    preparation: dict[str, Any]
    normalized_text: str
    diagnostics: dict[str, Any]


def _error(code: str, message: str) -> dict[str, Any]:
    return {"ok": False, "code": code, "message": message}


def _reject_json_constant(token: str) -> Any:
    raise ValueError(f"non-finite JSON number: {token}")


def _json_bytes(value: Any) -> bytes:
    return json.dumps(
        value,
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
        allow_nan=False,
    ).encode("utf-8")


def _max_depth(value: Any) -> int:
    if not isinstance(value, (dict, list)) or not value:
        return 1
    maximum = 1
    stack: list[tuple[Any, int]] = [(value, 1)]
    while stack:
        current, depth = stack.pop()
        maximum = max(maximum, depth)
        if depth > RAG_DOCUMENT_MAX_NESTING:
            return depth
        children = current.values() if isinstance(current, dict) else current
        for child in children:
            if isinstance(child, (dict, list)):
                stack.append((child, depth + 1))
    return maximum


def _normalized_key(value: Any) -> str:
    return "".join(character for character in str(value or "").lower() if character.isalnum())


def _sensitive_key(value: Any) -> bool:
    normalized = _normalized_key(value)
    return normalized.endswith("path") or any(part in normalized for part in _SENSITIVE_KEY_PARTS)


def normalize_document_text(value: str) -> str:
    """Match ``files.normalize_extracted_text`` exactly."""
    text = value.replace("\r\n", "\n").replace("\r", "\n").replace("\x00", "")
    lines = [line.rstrip() for line in text.split("\n")]
    return "\n".join(lines).strip()


def chunk_content_hash(text: str) -> str:
    """Match the existing Local RAG BLAKE2b-96 chunk lineage hash."""
    return hashlib.blake2b(str(text or "").strip().encode("utf-8", errors="ignore"), digest_size=12).hexdigest()


def document_content_hash(chunk_hash_by_index: dict[int, str]) -> str:
    """Match the existing Local RAG content-addressed document version."""
    digest = hashlib.blake2b(digest_size=12)
    for index in sorted(chunk_hash_by_index):
        digest.update(f"{index}:{chunk_hash_by_index[index]}\0".encode("utf-8", errors="ignore"))
    return digest.hexdigest()


def _normalize_metadata(value: Any) -> tuple[dict[str, Any] | None, dict[str, Any] | None]:
    if value is None:
        return {}, None
    if not isinstance(value, dict):
        return None, _error("invalid_metadata", "metadata must be an object")
    for key in value:
        if _sensitive_key(key):
            return None, _error("invalid_metadata", "metadata contains a path or credential field")
    normalized: dict[str, Any] = {}
    for key in sorted(_ALLOWED_METADATA):
        if key not in value:
            continue
        item = value[key]
        if item is None or isinstance(item, (str, bool, int)):
            normalized[key] = item
        elif isinstance(item, float) and math.isfinite(item):
            normalized[key] = item
        else:
            return None, _error("invalid_metadata", f"metadata.{key} must be a JSON scalar")
    return normalized, None


def _integer(value: Any) -> int | None:
    return value if isinstance(value, int) and not isinstance(value, bool) else None


def _chunk_text(
    text: str,
    *,
    document_id: str,
    chunk_chars: int,
    chunk_overlap: int,
) -> list[dict[str, Any]]:
    chunks: list[dict[str, Any]] = []
    start = 0
    text_length = len(text)
    while start < text_length:
        end = min(start + chunk_chars, text_length)
        if end < text_length:
            boundary = max(text.rfind("\n\n", start, end), text.rfind("\n", start, end))
            if boundary > start + chunk_chars // 2:
                end = boundary
        chunk_body = text[start:end].strip()
        if chunk_body:
            content_hash = chunk_content_hash(chunk_body)
            index = len(chunks)
            chunks.append(
                {
                    "index": index,
                    "chunkId": f"{document_id}:{index}:{content_hash}",
                    "text": chunk_body,
                    "start": start,
                    "end": end,
                    "lineStart": text.count("\n", 0, start) + 1,
                    "lineEnd": text.count("\n", 0, end) + 1,
                    "contentHash": content_hash,
                }
            )
        if end >= text_length:
            break
        start = max(end - chunk_overlap, start + 1)
    return chunks


def prepare_rag_document(
    value: Any,
    *,
    payload_size: int | None = None,
) -> dict[str, Any]:
    """Return a normalized document descriptor or a stable error object."""
    try:
        encoded = _json_bytes(value)
    except (TypeError, ValueError):
        return _error("invalid_request", "request must be safely JSON serializable")
    size = len(encoded) if payload_size is None else max(0, int(payload_size))
    if size > RAG_DOCUMENT_MAX_REQUEST_BYTES:
        return _error("request_too_large", "document preparation request is too large")
    if _max_depth(value) > RAG_DOCUMENT_MAX_NESTING:
        return _error("nesting_limit_exceeded", "document preparation request is too deeply nested")
    if not isinstance(value, dict):
        return _error("invalid_request", "request must be a JSON object")
    if any(_sensitive_key(key) for key in value):
        return _error("invalid_request", "request contains a path or credential field")
    if set(value) - _ALLOWED_TOP_LEVEL:
        return _error("invalid_request", "request contains unsupported fields")

    document_id = value.get("documentId")
    if (
        not isinstance(document_id, str)
        or not document_id
        or document_id != document_id.strip()
        or len(document_id) > RAG_DOCUMENT_MAX_ID_CHARACTERS
        or any(ord(character) < 32 for character in document_id)
    ):
        return _error("invalid_document_id", "documentId must be a normalized non-empty string")

    raw_text = value.get("text")
    if not isinstance(raw_text, str):
        return _error("invalid_text", "text must be a string")
    normalized_text = normalize_document_text(raw_text)
    if not normalized_text:
        return _error("invalid_text", "text must contain readable content")
    if len(normalized_text) > RAG_DOCUMENT_MAX_CHARACTERS:
        return _error("document_too_large", "normalized document exceeds the character limit")

    metadata, metadata_error = _normalize_metadata(value.get("metadata", {}))
    if metadata_error is not None:
        return metadata_error
    assert metadata is not None

    chunking = value.get("chunking")
    if not isinstance(chunking, dict):
        return _error("invalid_request", "chunking must be an object")
    if any(_sensitive_key(key) for key in chunking) or set(chunking) - _ALLOWED_CHUNKING:
        return _error("invalid_request", "chunking contains unsupported fields")
    chunk_chars = _integer(chunking.get("chunkChars"))
    if chunk_chars is None or chunk_chars <= 0 or chunk_chars > RAG_DOCUMENT_MAX_CHUNK_CHARACTERS:
        return _error("invalid_chunk_size", "chunkChars must be a positive bounded integer")
    chunk_overlap = _integer(chunking.get("chunkOverlap"))
    if chunk_overlap is None or chunk_overlap < 0:
        return _error("invalid_chunk_overlap", "chunkOverlap must be a non-negative integer")
    if chunk_overlap >= chunk_chars:
        return _error("chunk_overlap_too_large", "chunkOverlap must be smaller than chunkChars")

    chunks = _chunk_text(
        normalized_text,
        document_id=document_id,
        chunk_chars=chunk_chars,
        chunk_overlap=chunk_overlap,
    )
    hash_by_index = {int(chunk["index"]): str(chunk["contentHash"]) for chunk in chunks}
    return {
        "ok": True,
        "document": {
            "documentId": document_id,
            "contentHash": document_content_hash(hash_by_index),
            "characterCount": len(normalized_text),
            "chunkCount": len(chunks),
            "metadata": metadata,
        },
        "chunks": chunks,
        "chunking": {"chunkChars": chunk_chars, "chunkOverlap": chunk_overlap},
        "diagnostics": {"normalized": normalized_text != raw_text},
    }


def prepare_rag_document_json(raw: str | bytes) -> dict[str, Any]:
    encoded = raw.encode("utf-8") if isinstance(raw, str) else bytes(raw)
    if len(encoded) > RAG_DOCUMENT_MAX_REQUEST_BYTES:
        return _error("request_too_large", "document preparation request is too large")
    try:
        value = json.loads(encoded.decode("utf-8"), parse_constant=_reject_json_constant)
    except (UnicodeDecodeError, json.JSONDecodeError, ValueError):
        return _error("invalid_request", "request must contain valid JSON")
    return prepare_rag_document(value, payload_size=len(encoded))


def _diagnostics(value: dict[str, Any], local: dict[str, Any]) -> dict[str, Any]:
    document_id = value.get("documentId") if isinstance(value.get("documentId"), str) else ""
    chunking_value = value.get("chunking")
    chunking: dict[str, Any] = chunking_value if isinstance(chunking_value, dict) else {}
    document_value = local.get("document")
    document: dict[str, Any] = document_value if isinstance(document_value, dict) else {}
    return {
        "documentIdHash": hashlib.sha256(document_id.encode("utf-8", errors="ignore")).hexdigest()[:16] if document_id else "",
        "characterCount": int(document.get("characterCount") or 0),
        "chunkCount": int(document.get("chunkCount") or 0),
        "chunkSize": int(chunking.get("chunkChars") or 0),
        "chunkOverlap": int(chunking.get("chunkOverlap") or 0),
        "runtime": "python",
        "fallback": False,
        "fallbackReason": "",
        "latencyMs": 0,
    }


def _contains_sensitive_key(value: Any) -> bool:
    stack = [value]
    while stack:
        current = stack.pop()
        if isinstance(current, dict):
            if any(_sensitive_key(key) for key in current):
                return True
            stack.extend(current.values())
        elif isinstance(current, list):
            stack.extend(current)
    return False


def _validate_rust_success(local: dict[str, Any], candidate: Any, normalized_text: str) -> tuple[bool, str]:
    if not isinstance(candidate, dict):
        return False, "rust_response_not_object"
    try:
        _json_bytes(candidate)
    except (TypeError, ValueError):
        return False, "rust_response_not_serializable"
    if candidate.get("ok") is not True:
        return False, "rust_contract_invalid"
    if _contains_sensitive_key(candidate):
        return False, "rust_sensitive_field_injected"
    document = candidate.get("document")
    chunks = candidate.get("chunks")
    if not isinstance(document, dict) or not isinstance(chunks, list):
        return False, "rust_contract_invalid"
    local_document_value = local.get("document")
    local_document: dict[str, Any] = local_document_value if isinstance(local_document_value, dict) else {}
    if document.get("documentId") != local_document.get("documentId"):
        return False, "rust_document_id_changed"
    seen: set[str] = set()
    for expected_index, chunk in enumerate(chunks):
        if not isinstance(chunk, dict):
            return False, "rust_chunk_invalid"
        if chunk.get("index") != expected_index:
            return False, "rust_chunk_index_invalid"
        chunk_id = chunk.get("chunkId")
        if not isinstance(chunk_id, str) or not chunk_id or chunk_id in seen:
            return False, "rust_chunk_id_invalid"
        seen.add(chunk_id)
        start = chunk.get("start")
        end = chunk.get("end")
        if (
            not isinstance(start, int)
            or isinstance(start, bool)
            or not isinstance(end, int)
            or isinstance(end, bool)
            or not 0 <= start <= end <= len(normalized_text)
        ):
            return False, "rust_chunk_offset_invalid"
        if chunk.get("text") != normalized_text[start:end].strip():
            return False, "rust_chunk_text_mismatch"
        content_hash = chunk.get("contentHash")
        if not isinstance(content_hash, str) or len(content_hash) != 24 or any(c not in "0123456789abcdef" for c in content_hash):
            return False, "rust_content_hash_invalid"
        if content_hash != chunk_content_hash(str(chunk.get("text") or "")):
            return False, "rust_content_hash_mismatch"
    if candidate != local:
        return False, "rust_semantic_divergence"
    return True, ""


def rag_document_preparation_enabled() -> bool:
    value = os.environ.get("DEEPSEEK_RUST_RAG_DOCUMENT_PREP", "0")
    return value.strip().lower() in {"1", "true", "yes", "on"}


def prepare_rag_document_with_optional_rust(value: dict[str, Any]) -> RagDocumentPreparationDecision:
    local = prepare_rag_document(value)
    text_value = value.get("text")
    normalized_text = normalize_document_text(text_value) if isinstance(text_value, str) else ""
    diagnostics = _diagnostics(value, local)
    if local.get("ok") is not True or not rag_document_preparation_enabled():
        return RagDocumentPreparationDecision(local, normalized_text, diagnostics)

    from deepseek_infra.infra.rust_core import rag_client

    safe_payload = {
        "documentId": value.get("documentId"),
        "text": value.get("text"),
        "metadata": local.get("document", {}).get("metadata", {}),
        "chunking": local.get("chunking", {}),
    }
    started = time.perf_counter()
    result = rag_client.prepare_document(safe_payload)
    diagnostics["latencyMs"] = max(result.latency_ms, int((time.perf_counter() - started) * 1000))
    if not result.ok:
        diagnostics.update(runtime="python", fallback=True, fallbackReason=result.error_kind or "rust_backend_unavailable")
        return RagDocumentPreparationDecision(local, normalized_text, diagnostics)
    valid, reason = _validate_rust_success(local, result.body, normalized_text)
    if not valid:
        diagnostics.update(runtime="python", fallback=True, fallbackReason=reason)
        return RagDocumentPreparationDecision(local, normalized_text, diagnostics)
    diagnostics.update(runtime="rust", fallback=False, fallbackReason="")
    return RagDocumentPreparationDecision(result.body, normalized_text, diagnostics)


def log_rag_document_preparation(diagnostics: dict[str, Any]) -> None:
    logger.info(
        "rag_document_preparation",
        extra={
            "document_id_hash": str(diagnostics.get("documentIdHash") or ""),
            "character_count": int(diagnostics.get("characterCount") or 0),
            "chunk_count": int(diagnostics.get("chunkCount") or 0),
            "chunk_size": int(diagnostics.get("chunkSize") or 0),
            "chunk_overlap": int(diagnostics.get("chunkOverlap") or 0),
            "runtime": str(diagnostics.get("runtime") or "python"),
            "fallback": bool(diagnostics.get("fallback")),
            "fallback_reason": str(diagnostics.get("fallbackReason") or ""),
            "latency_ms": int(diagnostics.get("latencyMs") or 0),
        },
    )


def public_rag_document_diagnostics(diagnostics: dict[str, Any]) -> dict[str, Any]:
    return {
        "runtime": str(diagnostics.get("runtime") or "python"),
        "fallback": bool(diagnostics.get("fallback")),
        "fallbackReason": str(diagnostics.get("fallbackReason") or ""),
        "characterCount": int(diagnostics.get("characterCount") or 0),
        "chunkCount": int(diagnostics.get("chunkCount") or 0),
        "latencyMs": int(diagnostics.get("latencyMs") or 0),
    }

File: rust/crates/deepseek-rag/src/document_preparation.rs
//! Side-effect-free preparation of text already parsed by Python.

use blake2::Blake2bVar;
use blake2::digest::{Update, VariableOutput};
use serde_json::{Map, Value, json};
use std::collections::BTreeMap;

pub const MAX_DOCUMENT_CHARACTERS: usize = 8_000_000;
pub const MAX_REQUEST_BYTES: usize = 40_000_000;
pub const MAX_NESTING: usize = 24;
pub const MAX_DOCUMENT_ID_CHARACTERS: usize = 256;
pub const MAX_CHUNK_CHARACTERS: usize = 1_000_000;

const TOP_LEVEL_KEYS: &[&str] = &["documentId", "text", "metadata", "chunking"];
const CHUNKING_KEYS: &[&str] = &["chunkChars", "chunkOverlap"];
const METADATA_KEYS: &[&str] = &["displayName", "sourceType", "kind"];
const SENSITIVE_KEY_PARTS: &[&str] = &[
    "absolutepath",
    "temporarypath",
    "uploadpath",
    "cachepath",
    "authorization",
    "apikey",
    "token",
    "rawfilebytes",
    "databaselocation",
    "workspacesecret",
    "filesystempath",
];

fn error(code: &'static str, message: &'static str) -> Value {
    json!({"ok": false, "code": code, "message": message})
}

fn normalized_key(value: &str) -> String {
    value
        .chars()
        .flat_map(char::to_lowercase)
        .filter(|character| character.is_alphanumeric())
        .collect()
}

fn sensitive_key(value: &str) -> bool {
    let normalized = normalized_key(value);
    normalized.ends_with("path")
        || SENSITIVE_KEY_PARTS
            .iter()
            .any(|part| normalized.contains(part))
}

fn max_depth(value: &Value) -> usize {
    let mut maximum = 1;
    let mut stack = vec![(value, 1usize)];
    while let Some((current, depth)) = stack.pop() {
        maximum = maximum.max(depth);
        if depth > MAX_NESTING {
            return depth;
        }
        match current {
            Value::Array(items) => {
                for item in items {
                    if item.is_array() || item.is_object() {
                        stack.push((item, depth + 1));
                    }
                }
            }
            Value::Object(items) => {
                for item in items.values() {
                    if item.is_array() || item.is_object() {
                        stack.push((item, depth + 1));
                    }
                }
            }
            _ => {}
        }
    }
    maximum
}

pub fn normalize_document_text(value: &str) -> String {
    let line_endings = value
        .replace("\r\n", "\n")
        .replace('\r', "\n")
        .replace('\0', "");
    line_endings
        .split('\n')
        .map(str::trim_end)
        .collect::<Vec<_>>()
        .join("\n")
        .trim()
        .to_string()
}

fn blake2b_96(bytes: &[u8]) -> String {
    let mut hasher = Blake2bVar::new(12).expect("BLAKE2b-96 output size is valid");
    hasher.update(bytes);
    let mut output = [0u8; 12];
    hasher
        .finalize_variable(&mut output)
        .expect("fixed output buffer has the configured size");
    output.iter().map(|byte| format!("{byte:02x}")).collect()
}

pub fn chunk_content_hash(text: &str) -> String {
    blake2b_96(text.trim().as_bytes())
}

pub fn document_content_hash(chunk_hashes: &BTreeMap<usize, String>) -> String {
    let mut bytes = Vec::new();
    for (index, content_hash) in chunk_hashes {
        bytes.extend_from_slice(format!("{index}:{content_hash}\0").as_bytes());
    }
    blake2b_96(&bytes)
}

fn normalize_metadata(value: Option<&Value>) -> Result<Map<String, Value>, Value> {
    let Some(value) = value else {
        return Ok(Map::new());
    };
    let Some(metadata) = value.as_object() else {
        return Err(error("invalid_metadata", "metadata must be an object"));
    };
    if metadata.keys().any(|key| sensitive_key(key)) {
        return Err(error(
            "invalid_metadata",
            "metadata contains a path or credential field",
        ));
    }
    let mut normalized = Map::new();
    for key in METADATA_KEYS {
        let Some(item) = metadata.get(*key) else {
            continue;
        };
        if item.is_null() || item.is_string() || item.is_boolean() || item.is_number() {
            normalized.insert((*key).to_string(), item.clone());
        } else {
            return Err(error(
                "invalid_metadata",
                "allowlisted metadata values must be JSON scalars",
            ));
        }
    }
    Ok(normalized)
}

fn bounded_integer(value: Option<&Value>) -> Option<i64> {
    value.and_then(Value::as_i64)
}

fn chunk_text(
    text: &str,
    document_id: &str,
    chunk_chars: usize,
    chunk_overlap: usize,
) -> Vec<Value> {
    let characters: Vec<char> = text.chars().collect();
    let text_length = characters.len();
    let mut chunks = Vec::new();
    let mut start = 0usize;
    while start < text_length {
        let mut end = (start + chunk_chars).min(text_length);
        if end < text_length {
            let boundary = (start..end).rev().find(|index| characters[*index] == '\n');
            if let Some(boundary) = boundary
                && boundary > start + chunk_chars / 2
            {
                end = boundary;
            }
        }
        let raw: String = characters[start..end].iter().collect();
        let body = raw.trim().to_string();
        if !body.is_empty() {
            let index = chunks.len();
            let content_hash = chunk_content_hash(&body);
            let line_start = characters[..start]
                .iter()
                .filter(|character| **character == '\n')
                .count()
                + 1;
            let line_end = characters[..end]
                .iter()
                .filter(|character| **character == '\n')
                .count()
                + 1;
            chunks.push(json!({
                "index": index,
                "chunkId": format!("{document_id}:{index}:{content_hash}"),
                "text": body,
                "start": start,
                "end": end,
                "lineStart": line_start,
                "lineEnd": line_end,
                "contentHash": content_hash,
            }));
        }
        if end >= text_length {
            break;
        }
        start = end.saturating_sub(chunk_overlap).max(start + 1);
    }
    chunks
}

pub fn prepare_document_value(value: &Value, payload_size: usize) -> Value {
    if payload_size > MAX_REQUEST_BYTES {
        return error(
            "request_too_large",
            "document preparation request is too large",
        );
    }
    if max_depth(value) > MAX_NESTING {
        return error(
            "nesting_limit_exceeded",
            "document preparation request is too deeply nested",
        );
    }
    let Some(request) = value.as_object() else {
        return error("invalid_request", "request must be a JSON object");
    };
    if request.keys().any(|key| sensitive_key(key)) {
        return error(
            "invalid_request",
            "request contains a path or credential field",
        );
    }
    if request
        .keys()
        .any(|key| !TOP_LEVEL_KEYS.contains(&key.as_str()))
    {
        return error("invalid_request", "request contains unsupported fields");
    }

    let Some(document_id) = request.get("documentId").and_then(Value::as_str) else {
        return error(
            "invalid_document_id",
            "documentId must be a normalized non-empty string",
        );
    };
    if document_id.is_empty()
        || document_id.trim() != document_id
        || document_id.chars().count() > MAX_DOCUMENT_ID_CHARACTERS
        || document_id.chars().any(|character| character < ' ')
    {
        return error(
            "invalid_document_id",
            "documentId must be a normalized non-empty string",
        );
    }
    let Some(raw_text) = request.get("text").and_then(Value::as_str) else {
        return error("invalid_text", "text must be a string");
    };
    let normalized_text = normalize_document_text(raw_text);
    if normalized_text.is_empty() {
        return error("invalid_text", "text must contain readable content");
    }
    if normalized_text.chars().count() > MAX_DOCUMENT_CHARACTERS {
        return error(
            "document_too_large",
            "normalized document exceeds the character limit",
        );
    }
    let metadata = match normalize_metadata(request.get("metadata")) {
        Ok(metadata) => metadata,
        Err(error) => return error,
    };
    let Some(chunking) = request.get("chunking").and_then(Value::as_object) else {
        return error("invalid_request", "chunking must be an object");
    };
    if chunking.keys().any(|key| sensitive_key(key))
        || chunking
            .keys()
            .any(|key| !CHUNKING_KEYS.contains(&key.as_str()))
    {
        return error("invalid_request", "chunking contains unsupported fields");
    }
    let Some(chunk_chars) = bounded_integer(chunking.get("chunkChars")) else {
        return error(
            "invalid_chunk_size",
            "chunkChars must be a positive bounded integer",
        );
    };
    if chunk_chars <= 0 || chunk_chars as usize > MAX_CHUNK_CHARACTERS {
        return error(
            "invalid_chunk_size",
            "chunkChars must be a positive bounded integer",
        );
    }
    let Some(chunk_overlap) = bounded_integer(chunking.get("chunkOverlap")) else {
        return error(
            "invalid_chunk_overlap",
            "chunkOverlap must be a non-negative integer",
        );
    };
    if chunk_overlap < 0 {
        return error(
            "invalid_chunk_overlap",
            "chunkOverlap must be a non-negative integer",
        );
    }
    if chunk_overlap >= chunk_chars {
        return error(
            "chunk_overlap_too_large",
            "chunkOverlap must be smaller than chunkChars",
        );
    }

    let chunks = chunk_text(
        &normalized_text,
        document_id,
        chunk_chars as usize,
        chunk_overlap as usize,
    );
    let chunk_hashes = chunks
        .iter()
        .filter_map(|chunk| {
            Some((
                chunk.get("index")?.as_u64()? as usize,
                chunk.get("contentHash")?.as_str()?.to_string(),
            ))
        })
        .collect::<BTreeMap<_, _>>();
    json!({
        "ok": true,
        "document": {
            "documentId": document_id,
            "contentHash": document_content_hash(&chunk_hashes),
            "characterCount": normalized_text.chars().count(),
            "chunkCount": chunks.len(),
            "metadata": metadata,
        },
        "chunks": chunks,
        "chunking": {"chunkChars": chunk_chars, "chunkOverlap": chunk_overlap},
        "diagnostics": {"normalized": normalized_text != raw_text},
    })
}

pub fn prepare_document_bytes(body: &[u8]) -> Value {
    if body.len() > MAX_REQUEST_BYTES {
        return error(
            "request_too_large",
            "document preparation request is too large",
        );
    }
    let value: Value = match serde_json::from_slice(body) {
        Ok(value) => value,
        Err(_) => return error("invalid_request", "request must contain valid JSON"),
    };
    prepare_document_value(&value, body.len())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn request(text: &str, chunk_chars: i64, chunk_overlap: i64) -> Value {
        json!({
            "documentId":"doc-1",
            "text":text,
            "metadata":{"displayName":"notes.txt","sourceType":"text/plain"},
            "chunking":{"chunkChars":chunk_chars,"chunkOverlap":chunk_overlap}
        })
    }

    #[test]
    fn prepares_minimal_document() {
        let result = prepare_document_value(&request("hello", 6000, 400), 100);
        assert_eq!(result["ok"], true);
        assert_eq!(result["chunks"][0]["text"], "hello");
    }

    #[test]
    fn normalizes_line_endings() {
        let result = prepare_document_value(&request(" a  \r\n b\r\n", 20, 0), 100);
        assert_eq!(result["chunks"][0]["text"], "a\n b");
        assert_eq!(result["document"]["characterCount"], 4);
    }

    #[test]
    fn chunks_exact_boundary() {
        let result = prepare_document_value(&request("abcdefgh", 4, 0), 100);
        assert_eq!(result["document"]["chunkCount"], 2);
        assert_eq!(result["chunks"][1]["start"], 4);
    }

    #[test]
    fn chunks_with_overlap() {
        let result = prepare_document_value(&request("abcdefghij", 6, 2), 100);
        assert_eq!(result["chunks"][0]["text"], "abcdef");
        assert_eq!(result["chunks"][1]["text"], "efghij");
    }

    #[test]
    fn prevents_overlap_infinite_loop() {
        let result = prepare_document_value(&request("abcdefgh", 4, 4), 100);
        assert_eq!(result["code"], "chunk_overlap_too_large");

        let boundary = prepare_document_value(&request("aaaaaa\nbbbbbbbbbbbb", 10, 9), 100);
        assert_eq!(boundary["ok"], true);
        assert!(boundary["chunks"].as_array().unwrap().len() <= 19);
    }

    #[test]
    fn preserves_paragraph_boundaries() {
        let result = prepare_document_value(&request("aaaa\n\nbbbb\ncccc", 12, 0), 100);
        assert_eq!(result["chunks"][0]["text"], "aaaa\n\nbbbb");
    }

    #[test]
    fn preserves_cjk() {
        let result = prepare_document_value(&request("中文分块测试", 3, 1), 100);
        assert_eq!(result["chunks"][0]["text"], "中文分");
        assert_eq!(result["chunks"][1]["start"], 2);
    }

    #[test]
    fn preserves_emoji() {
        let result = prepare_document_value(&request("A🚀B🙂C", 3, 1), 100);
        assert_eq!(result["chunks"][0]["text"], "A🚀B");
    }

    #[test]
    fn uses_character_offsets_not_byte_offsets() {
        let result = prepare_document_value(&request("中🚀ab", 2, 0), 100);
        assert_eq!(result["chunks"][1]["start"], 2);
        assert_eq!(result["chunks"][1]["text"], "ab");
    }

    #[test]
    fn handles_combining_characters() {
        let result = prepare_document_value(&request("e\u{301}x", 2, 0), 100);
        assert_eq!(result["chunks"][0]["end"], 2);
        assert_eq!(result["chunks"][0]["text"], "e\u{301}");
    }

    #[test]
    fn generates_stable_document_hash() {
        let left = prepare_document_value(&request("stable text", 6, 1), 100);
        let right = prepare_document_value(&request("stable text", 6, 1), 100);
        assert_eq!(
            left["document"]["contentHash"],
            right["document"]["contentHash"]
        );
    }

    #[test]
    fn generates_unique_chunk_ids() {
        let result = prepare_document_value(&request("abcdefghijkl", 4, 1), 100);
        let ids = result["chunks"]
            .as_array()
            .unwrap()
            .iter()
            .map(|chunk| chunk["chunkId"].as_str().unwrap())
            .collect::<std::collections::HashSet<_>>();
        assert_eq!(ids.len(), result["chunks"].as_array().unwrap().len());
    }

    #[test]
    fn rejects_invalid_chunk_size() {
        assert_eq!(
            prepare_document_value(&request("x", 0, 0), 100)["code"],
            "invalid_chunk_size"
        );
    }

    #[test]
    fn rejects_invalid_overlap() {
        assert_eq!(
            prepare_document_value(&request("x", 4, -1), 100)["code"],
            "invalid_chunk_overlap"
        );
    }

    #[test]
    fn rejects_oversized_document() {
        let value = request("small", 4, 0);
        assert_eq!(
            prepare_document_value(&value, MAX_REQUEST_BYTES + 1)["code"],
            "request_too_large"
        );
        assert_eq!(MAX_DOCUMENT_CHARACTERS, 8_000_000);
    }

    #[test]
    fn rejects_excessive_nesting() {
        let mut nested = json!({});
        for _ in 0..=MAX_NESTING {
            nested = json!({"nested": nested});
        }
        let mut value = request("x", 4, 0);
        value["metadata"] = nested;
        assert_eq!(
            prepare_document_value(&value, 100)["code"],
            "nesting_limit_exceeded"
        );
    }

    #[test]
    fn rejects_path_like_sensitive_fields() {
        let mut value = request("x", 4, 0);
        value["metadata"]["absolutePath"] = json!("C:/secret.txt");
        assert_eq!(
            prepare_document_value(&value, 100)["code"],
            "invalid_metadata"
        );
    }

    #[test]
    fn response_roundtrips_json() {
        let result = prepare_document_value(&request("中文🚀", 2, 0), 100);
        let encoded = serde_json::to_vec(&result).unwrap();
        assert_eq!(serde_json::from_slice::<Value>(&encoded).unwrap(), result);
    }

    #[test]
    fn never_reads_files() {
        let mut value = request("already parsed", 20, 0);
        value["uploadPath"] = json!("/tmp/upload");
        assert_eq!(
            prepare_document_value(&value, 100)["code"],
            "invalid_request"
        );
    }

    #[test]
    fn never_writes_index() {
        let value = request("pure transformation", 20, 0);
        let first = prepare_document_value(&value, 100);
        let second = prepare_document_value(&value, 100);
        assert_eq!(first, second);
    }

    #[test]
    fn never_receives_credentials() {
        let mut value = request("x", 4, 0);
        value["metadata"]["authorization"] = json!("Bearer secret");
        assert_eq!(
            prepare_document_value(&value, 100)["code"],
            "invalid_metadata"
        );
    }
}
